File: src/FileLogger.Old/FileLogger.cs
using Microsoft.Extensions.Logging;

namespace FileLogger.Old;

public sealed class FileLogger(
    string category,
    Func<FileLoggerConfiguration> getCurrentConfig,
    ILogWriter logWriter) : ILogger
{
    public IDisposable? BeginScope<TState>(TState state) where TState : notnull => default!;

    public bool IsEnabled(LogLevel logLevel) => logLevel >= getCurrentConfig().MinLevel;

    public void Log<TState>(
        LogLevel logLevel,
        EventId eventId,
        TState state,
        Exception? exception,
        Func<TState, Exception?, string> formatter)
    {
        if (!IsEnabled(logLevel))
        {
            return;
        }
        
        var timeStamp = $"{DateTime.Now:yyyy-MM-dd HH:mm:ss.fff}";
        var threadId = $"{Environment.CurrentManagedThreadId:##}";
        var level = $"{logLevel,-5}";
        var message = formatter(state, exception);

        if (exception != null)
        {
            message = message + Environment.NewLine + exception;
        }
        
        var logEntry = $"{timeStamp} [{threadId}] [{level}] [{category}] {message}";
        logWriter.EnqueueWrite(logEntry);
    }
}

File: src/FileLogger/FileLogger.fs
namespace FileLogger

open System
open System.Runtime.CompilerServices
open Microsoft.Extensions.Logging

type FileLogger(category: string, timeProvider: TimeProvider, writeLog) =
    interface ILogger with
        member this.BeginScope(state) = Unsafe.NullRef<IDisposable>()
        member this.IsEnabled(logLevel) = true //TODO: respect logLevel

        member this.Log(logLevel, eventId, state, except, formatter) =
            let logEntry =
                { Time = timeProvider.GetLocalNow()
                  ThreadId = Environment.CurrentManagedThreadId
                  Level = logLevel
                  Category = category
                  Message = formatter.Invoke(state, except)
                  Exception = Option.ofObj except }

            writeLog logEntry
